File: lib/iriq/normalizer.rb
module Iriq
  # Produces a canonical, shape-aware string for an identifier.
  #
  #   Normalizer.normalize("https://Foo.com:443/users/123")
  #   # => "https://foo.com/users/{user_id}"
  #
  # The form is intended for grouping/diffing — it is not a round-trippable URL.
  #
  # Path + query rendering dispatches through an evidence source so the
  # mechanical (classifier-only) and corpus-informed code paths share one
  # entry point. When `evidence` is nil, NullEvidenceSource provides the
  # mechanical behavior (PathShape + param-name-hint query rules). When a
  # Corpus is passed as `evidence`, its observed Position / Cluster stats
  # drive the rendering (variability promotion, popular outlier
  # preservation, cluster-inferred query types).
  module Normalizer
    module_function

    def normalize(input, classifier: SegmentClassifier::DEFAULT, hints: true, evidence: nil)
      iri = input.is_a?(Identifier) ? input : Parser.parse(input)
      normalize_identifier(iri, classifier: classifier, hints: hints, evidence: evidence)
    end

    def normalize_identifier(iri, classifier: SegmentClassifier::DEFAULT, hints: true, evidence: nil)
      return normalize_urn(iri, classifier, hints) if iri.urn?

      src = evidence || NullEvidenceSource.new
      out = +""
      out << "#{iri.scheme}://" if iri.scheme
      out << iri.host if iri.host
      out << ":#{iri.port}" if iri.port
      out << src.render_path(iri, classifier, hints)
      if iri.query_params && !iri.query_params.empty?
        out << "?" << src.render_query(iri, classifier)
      end
      out
    end

    def normalize_urn(iri, classifier, hints)
      return iri.canonical unless iri.scheme == "urn" && iri.nss && iri.nss.include?(":")

      ns, value = iri.nss.split(":", 2)
      entry     = SegmentHints.derive([ns, value], classifier).last
      shaped =
        if entry[:type] == :date && (canon = SegmentClassifier.canonical_date(entry[:value]))
          canon
        elsif entry[:type] == :currency && (canon = SegmentClassifier.canonical_currency(entry[:value]))
          canon
        elsif entry[:variable]
          "{#{(hints && entry[:hint]) || SegmentClassifier.display_type(entry[:type])}}"
        else
          entry[:value]
        end
      "urn:#{ns}:#{shaped}"
    end
  end

  # NullEvidenceSource is the default evidence source — purely
  # classifier-driven, no corpus signal. The Normalizer's mechanical
  # behavior is what this produces. Implements the same {render_path,
  # render_query} interface that Corpus implements for the corpus-informed
  # path.
  class NullEvidenceSource
    def render_path(iri, classifier, hints)
      PathShape.new(
        classifier: classifier, hints: hints,
        canonical_dates: true, canonical_currencies: true,
      ).for(iri.path_segments)
    end

    def render_query(iri, classifier)
      iri.query_params.keys.sort.map do |k|
        v    = iri.query_params[k]
        type = classifier.classify(v.to_s)
        # Param-name hint can lift a generic literal/opaque_id/slug into
        # a semantic type — `?phone=unknown` becomes `{phone}`.
        if (hint = SegmentClassifier.param_name_hint(k, type))
          type = hint
        end
        shaped =
          if type == :date && (canon = SegmentClassifier.canonical_date(v.to_s))
            canon
          elsif type == :currency && (canon = SegmentClassifier.canonical_currency(v.to_s))
            canon
          elsif classifier.variable?(type)
            "{#{SegmentClassifier.display_type(type)}}"
          else
            v
          end
        "#{k}=#{shaped}"
      end.join("&")
    end
  end
end

File: rust/iriq/src/normalizer.rs
use crate::classifier::{
    canonical_currency, canonical_date, display_type, param_name_hint, SegmentClassifier,
    SegmentType, DEFAULT_CLASSIFIER,
};
use crate::errors::ParseError;
use crate::hints::derive_hints;
use crate::identifier::Identifier;
use crate::parser::parse;
use crate::path_shape::PathShape;
use std::convert::Infallible;

pub trait NormalizationEvidence {
    /// What a failed evidence read reports.
    type Error;
    fn render_path(
        &self,
        iri: &Identifier,
        c: &SegmentClassifier,
        hints: bool,
    ) -> Result<String, Self::Error>;
    fn render_query(&self, iri: &Identifier, c: &SegmentClassifier) -> Result<String, Self::Error>;
}

pub struct NullEvidence;

impl NormalizationEvidence for NullEvidence {
    type Error = Infallible;

    fn render_path(
        &self,
        iri: &Identifier,
        c: &SegmentClassifier,
        hints: bool,
    ) -> Result<String, Infallible> {
        let mut ps = PathShape::new();
        ps.classifier = c;
        ps.hints = hints;
        ps.canonical_dates = true;
        ps.canonical_currencies = true;
        Ok(ps.for_segments(&iri.path_segments))
    }

    fn render_query(&self, iri: &Identifier, c: &SegmentClassifier) -> Result<String, Infallible> {
        Ok(shape_query(iri, c))
    }
}

pub fn normalize(input: &str) -> Result<String, ParseError> {
    normalize_with(input, &DEFAULT_CLASSIFIER, true)
}

pub fn normalize_with(
    input: &str,
    c: &SegmentClassifier,
    hints: bool,
) -> Result<String, ParseError> {
    let iri = parse(input)?;
    Ok(normalize_identifier_with(&iri, c, hints))
}

/// Normalize an already-parsed IRI (see [`normalize`]). `hints: false`
/// renders bare type placeholders (`{integer}`) instead of hinted names
/// (`{user_id}`).
pub fn normalize_identifier(iri: &Identifier, hints: bool) -> String {
    normalize_identifier_with(iri, &DEFAULT_CLASSIFIER, hints)
}

pub(crate) fn normalize_identifier_with(
    iri: &Identifier,
    c: &SegmentClassifier,
    hints: bool,
) -> String {
    let Ok(s) = normalize_identifier_with_evidence(iri, c, hints, &NullEvidence);
    s
}

pub fn normalize_identifier_with_evidence<E: NormalizationEvidence>(
    iri: &Identifier,
    c: &SegmentClassifier,
    hints: bool,
    ev: &E,
) -> Result<String, E::Error> {
    if iri.is_urn() {
        return Ok(normalize_urn(iri, c, hints));
    }
    let mut s = String::new();
    if !iri.scheme.is_empty() {
        s.push_str(&iri.scheme);
        s.push_str("://");
    }
    if !iri.host.is_empty() {
        s.push_str(&iri.host);
    }
    if let Some(port) = iri.port {
        s.push(':');
        s.push_str(&port.to_string());
    }
    s.push_str(&ev.render_path(iri, c, hints)?);
    if !iri.query_params.is_empty() {
        s.push('?');
        s.push_str(&ev.render_query(iri, c)?);
    }
    Ok(s)
}

fn normalize_urn(iri: &Identifier, c: &SegmentClassifier, hints: bool) -> String {
    if !(iri.scheme == "urn" && !iri.nss.is_empty() && iri.nss.contains(':')) {
        return iri.canonical();
    }
    let (ns, value) = iri.nss.split_once(':').unwrap();
    let entries = derive_hints(&[ns.to_string(), value.to_string()], c);
    let entry = entries.last().unwrap();
    let shaped = if entry.ty == SegmentType::Date {
        if let Some(canon) = canonical_date(&entry.value) {
            canon
        } else {
            placeholder(entry, hints)
        }
    } else if entry.ty == SegmentType::Currency {
        if let Some(canon) = canonical_currency(&entry.value) {
            canon
        } else {
            placeholder(entry, hints)
        }
    } else if entry.variable {
        placeholder(entry, hints)
    } else {
        entry.value.clone()
    };
    format!("urn:{}:{}", ns, shaped)
}

fn placeholder(entry: &crate::hints::SegmentHint, hints: bool) -> String {
    let p = if hints && !entry.hint.is_empty() {
        entry.hint.clone()
    } else {
        display_type(&entry.ty).to_string()
    };
    format!("{{{}}}", p)
}

fn shape_query(iri: &Identifier, c: &SegmentClassifier) -> String {
    let mut keys = iri.query_params.keys();
    keys.sort();
    let mut parts: Vec<String> = Vec::with_capacity(keys.len());
    for k in keys {
        let v = iri.query_params.get(&k).unwrap_or("");
        parts.push(format!("{}={}", k, render_param(&k, v, c)));
    }
    parts.join("&")
}

/// One param's mechanical rendering. A corpus renders a param it lacks
/// evidence on through this, so the two paths can't drift.
pub(crate) fn render_param(name: &str, value: &str, c: &SegmentClassifier) -> String {
    let mut t = c.classify(value);
    if let Some(h) = param_name_hint(name, &t) {
        t = h;
    }
    let canon = match t {
        SegmentType::Date => canonical_date(value),
        SegmentType::Currency => canonical_currency(value),
        _ => None,
    };
    match canon {
        Some(canon) => canon,
        None if matches!(t, SegmentType::Date | SegmentType::Currency) || c.variable(&t) => {
            format!("{{{}}}", display_type(&t))
        }
        None => value.to_string(),
    }
}
